### PR TITLE
Richer array literal element type inference

### LINKED ISSUE #8196: Determine eltType from "most general type" for all array elements
It appears I have to take all my integers in an array and put either a `.0` or at least `.` after the number.  For instance

```chapel
var d = Matrix([3.0 , 7, 4.3], [4.7 , 9, 22.8]);
> pginsert.chpl:29: error: Array literal element 2 expected to be of type real(64) but is of type int(64)
```

I think Chapel arrays assume `reals`, so why not just take that 7 at `7.0`?

### PR DESCRIPTION
**Pull request by bradcray**
In #8196, users have (understandably) complained about how poor
Chapel's type unification for array literals is.  Specifically, an array
literal's type is currently based on its first element's type, and all
subsequent elements must match precisely or an error is generated.
This means that `[1.0, 2]` and `[1, 2.0]` are inferred to be array of
real and array of int respectively, and that an error is flagged because
the second element doesn't match the initial one.

In this PR, I've implemented a module-based strategy to improve upon
this behavior.  With it, the module code will inspect all array elements,
trying to determine a most satisfying type for them.  I did this by writing
a utility routine that iterates over all of the elements in a param loop,
returning each of them in its own iteration, and then querying the return
type of that routine to leverage return type unification.  This causes the
arrays above to each be inferred to be "array of real" and for the integer
values to turn into their real equivalents.

A downside to this strategy is that it results in code (or, at least, AST),
that's proporational in size to the array literal in order to return all the
elements.  That said, so does the pre-existing machinery that builds
the array literal.

In order to generate a reasonable error message, I tagged my routine
with a pragma and made the compiler recognize it when complaining
about inability to resolve return types, substituting in an array-centric
message.

While here, I added special-case code for homogeneous array literals to
avoid the current and new param loops in order to reduce code size and
improve compilation times (and possibly execution times?).  I also noticed
that we had some ancient code that was still turning arrays of c-strings
into arrays of strings (and similarly for other data types), which is a remnant
of when we only had one string literal type and they were all C string literals.
That feels simply wrong in today's world where strings and C-strings have
their own literal formats, so as part of clean-up I took it out.

I added some new tests of various degrees of array literal and updated some
existing tests to reflect the new behavior and error message.

Note that as the dyno compiler comes online, we'd really like to move this
reasoning about array literal types out of the module code and into the
compiler, unifying it with other type unification needs like the return type
inference.  To that end, this can be considered a baby step in the right
direction until we get there that should help users out (while also
hopefully reducing compile time for homogeneous array literals).

Resolves #8196.

### CODE REVIEW

**Comment by bradcray** [comment]
@DanilaFe may be interested in this because it's a poor-person's way to do type inference in module code.
@riftEmber may be interested in it because it's a compelling use of param for loops, as we were discussing in OH last week.

**Comment by vasslitvinov** [comment]
The issue with getting the undesirable `USR_FATAL(fn, "unable to resolve return type")` can be dealt with by wrapping that in `if (inTryResolve == 0)`, like we do in a couple of places in functionResolution.cpp.

The question is what needs to happen if inTryResolve != 0. Ideally, not much. However, looking at the code, it was not obvious to me what the protocall is.

@mppf would you offer wisdome? What I am thinking is that in resolveReturnTypeAndYieldedType() we would have:
```c++
if (inTryResolve == 0) {
  USR_FATAL(fn, "unable to resolve return type");
} else {
  .... what should go here? ....
}
```

I confirmed that inTryResolve==1 in the case of a non-unifiable tuple passed into Brad's chpl_computeUnifiedType().

I am writing all this because I believe we should be unifying the element types of an array literal, rather than using its first element's type.

**Comment by mppf** [comment]
For the protocol for tryResolve -- see `tryResolveFunction`. Probably what goes in that `else` block is `tryResolveStates.back() = CHECK_FAILED;`.

**Comment by bradcray** [comment]
@vasslitvinov / @mppf: Thanks for taking at least a brief look.  Any opinions about the direction of this PR as a whole and/or whether you think it should or should not go into the release?  Thanks.

**Comment by vasslitvinov** [comment]
I would like this to go in the release, ideally with inferArrayLiteralTypeFromFirstElt defaulting to false or be removed altogether as if it were false. At least this is my intuition.

The changes in the PR look benign enough to me.

**Comment by ghbrown** [comment]
I'm posting my findings here since they're change/PR specific.

As described, I think both options are a step in the right direction, and I really appreciate the quick response and work on the issue. If the error message could be improved, I would be fine with having the current opt-in strategy be on by default until the hopefully more comprehensive fix. Specifically because I suspect that most arrays to which this would be applied are of quite modest size (unless these routines get called to read matrices from files, etc.) and apparently the code already scales proportional to array size per Brad.

The new default inference strategy gives
```chapel
var mat_0 = [[1.0,0,0],[0.0,0,0]]; // compiles without error (as expected)
var mat_1 = [[1.0,0,0],[0,0,0.0]]; // error about expecting int but got real (expected?)
var mat_2 = [[1.0,0,0],[0,0,0]]; // gives error about unresolved call due to different types of 1-D arrays (expected?)
```

**I couldn't get the opt-int strategy to work though**
```
error: Trying to set unrecognized config 'inferArrayLiteralFromFirstElt' via -s flag
```
I did check my version and it has the same hash as your branch, but didn't look much further than that.

After seeing the `mat_1` error where the default scheme fails on an order 2 array, I was going probe whether the opt-in scheme "properly" handled it (make whole array real using one element in array).

A generic version for any order array may be a bit more of a hassle to code (hopefully possible), but since different sub-arrays having different types is a direct route to a compiler error anyway I would personally want the default behavior to be: A[0,0,...,0] element sets array type.

**Summary: hopefully both schemes can be extended to infer types for arbitrary order arrays?**

**Comment by bradcray** [comment]
@ghbrown : Thanks for taking a look so fast.

W.r.t, option 2: oops, I had a typo on my PR's message.  The config should be `-sinferArrayLiteralTypeFromFirstElt=false`  I've updated the OP to correct this.

> I would be fine with having the current opt-in strategy be on by default until the hopefully more comprehensive fix.

I hear both you and Vass saying this, but I personally wouldn't be comfortable making it the default for this upcoming release.  Specifically, I worry about making compile times blow up without more testing and study, and if our nightly testing indicated that they do, I wouldn't have time to react to them before the release.  I _could_ see flipping the bit just after the release, though, with the hope of making it the default in the next release.  (Alternatively, we could save the whole thing for the next release, but I don't feel as nervous about the first strategy's impacts).

**Comment by ghbrown** [comment]
Here are the updated results from those tests. Seems the opt-in scheme doesn't catch higher order arrays either (wasn't expecting, just hoping and checking).
```chapel
var mat_0 = [[1.0,0,0],[0.0,0,0]]; // compiles without error (as expected)
var mat_1 = [[1.0,0,0],[0,0,0.0]]; // compiles without error (as expected)
var mat_2 = [[1.0,0,0],[0,0,0]]; // unable to resolve return type
```

Certainly no pressure from me to put anything specific in this release. I'd rather it be squashed thoroughly and logically, so even delaying everything to next release is no big deal as far as I'm concerned.

**Comment by bradcray** [comment]
Thanks for the quick look.  I wasn't certain what would happen in the `var mat_2 = [[1.0,0,0],[0,0,0]];` case, and am not completely surprised it didn't work, though I was cautiously optimistic.  That bears a bit more looking and thinking on my part to see if it's low-hanging fruit.

I suspect it would be easier to unify the types in the context of `Matrix([1.0,0,0],[0,0,0])`, though that's one of those clean-ups I mentioned needing to look into still (i.e., I'm almost certain it won't work correctly today).  Are you more interested in one of the Chapel-specific array-of-array cases, or the LinearAlgebra Matrix case?

I'll also note that we've long intended to support a "true" multidimensional array literal, whereas [[1, 2], [3, 4]] is a 1D array of 1D arrays rather than a Chapel 2D array, but have never quite gotten it done... yet.  If that's what you really want feel free to tag onto https://github.com/chapel-lang/chapel/issues/8864.

**Review by bmcdonald3 — APPROVED** [review]
I'm in favor of getting this into the release, especially given the bigger change is only provided as an opt-in feature with a long, obscure name.

My only concern would be the maintenance of the `inferArrayLiteralTypeFromFirstElt` param value. Would we rather rush this in to this release or wait until next release where we have reached a better idea of what we want the behavior to be and don't have to have this new, ugly param?

In my opinion, I'd just put it in now, given that it seems likely to fall by the wayside in the future.

**Comment by ghbrown** [comment]
> Are you more interested in one of the Chapel-specific array-of-array cases, or the LinearAlgebra Matrix case?

I think the LinearAlgebra Matrix case is more likely to have users inputting data in source files like my examples, and the true 2-D array it creates seems to be the more useful object.
On the other hand, I would guess the 1-D array of 1-D arrays gets used by new users before they know they can't slice them like true 2-D array.
So it's "fix actual use case" versus "protect new users" as I see it.

I should mention that what might be nice (no idea if it's possible) is that the eltType parameter of Matrix() controls how the elements are read, or at least sets the type that you cast into while reading. Right now it seems to do only whole arrays casts after reading.

> I'll also note that we've long intended to support a "true" multidimensional array literal

I think that's a worthwhile effort and would automatically protect the new users from arrays of arrays. I'll chime in there a bit later after I've caught up on the thread.

**Comment by bradcray** [comment]
Just to give people a read on how I'm feeling this morning after a late night:  Given that we're no worse off than we were in previous releases and that most people are clamoring for the more complete solution w/out config than the baby step forward I was proposing for this release, my current thinking is to hold this until after the release, merge a variant of the full-on version then, and check for compilation / performance impacts.

**Comment by bradcray** [comment]
@vasslitvinov / @mppf: I'm getting back to this PR in hopes of merging it after the development branch has thawed which means that I'm looking at your comments about improving the error message in earnest for the first time.  In doing so, I'm realizing that I don't really understand them.  I'm writing module code, and you seem to be pointing to compiler code / sketching out C++ code fragments rather than Chapel code.  What am I missing?  Is the idea that the compiler somehow detects that it is this specific case of being unable to resolve the return type, so it can specialize the error message?  And are there other pre-existing examples of this?  (in the two minutes I had before dinner, none were jumping out at me, but I'm admittedly moving fast).  Thanks.

**Comment by vasslitvinov** [comment]
@bradcray from OP:

> the error message is strictly worse in the event that the compiler can't unify the types (because it shows up as an "unable to resolve return type" style of error)

The compiler generates this error message at the point where you ask it `canResolve("chpl_computeUnifiedTypeHelp", x)` .  Here, you do not want the error message, you just want canResolve() to return `false`. The proposed compiler change aims at that.

At a high level, we proposed that the compiler treats "unable to resolve return type" just like a resolution error that occurs while resolving the target function's body.

**Comment by bradcray** [comment]
> At the same time, eventually, I'm expecting the dyno resolver to handle these cases more directly within the compiler. It will take some time before we are using that in production.

I agree (and said as much in the OP), but meanwhile, this is a pain point for real users, so if we can address it easily and replace it later, that seems like a worthwhile move to make.  As I understand it, it's still some time away (multiple releases?) before dyno is going to be ready to handle this case.

**Comment by bradcray** [comment]
I've cleaned up this PR to contain only the preferred "unify all array elements approach", and to specialize the error messages generated by the compiler.  @vasslitvinov or @bmcdonald3, since you've already given this at least a short look, would either of you be interested in, and able to, review it?  Or should I turn it loose to the group?  Let me know.

File: modules/internal/ChapelArray.chpl
Comment (vasslitvinov) [inline modules/internal/ChapelArray.chpl:325]: How about simply using `src.type` instead?

File: test/arrays/literals/hetLiteralsFirstGoverns.chpl
Comment (vasslitvinov) [inline test/arrays/literals/hetLiteralsFirstGoverns.chpl:5]: The name of this test suggests that "first governs" is a thing.

I suggest moving the contents of this test into `hetLiteralsUnifiable.chpl` and removing this test.

**Review by vasslitvinov — CHANGES_REQUESTED** [review]
Sounds good except for prim_assign. Please see the comments.

File: modules/internal/ChapelArray.chpl
Comment (vasslitvinov) [inline modules/internal/ChapelArray.chpl:312]: Perhaps computeUnifiedType() does not need to cater to the homogeneous case?

File: modules/internal/ChapelArray.chpl
Comment (vasslitvinov) [inline modules/internal/ChapelArray.chpl:335]: OK to keep this, however this should not happen, right? given that computeUnifyingType() completed successfully.  Perhaps we do not need the canResolve() check above at all?

**Review by vasslitvinov — APPROVED** [review]
Switching my review to approval, assuming follow-up on my comments.

Note I edited my comment on prim_assign.

File: compiler/resolution/resolveFunction.cpp
Comment (bradcray) [reply compiler/resolution/resolveFunction.cpp:1984]: I started out by trying to incorporate "try in reolve" logic, but since `inTryResolve` and the accompanying state variables are all in functionResolution.cpp rather than resolveFunction.cpp, it wasn't easy.

That said, I'm also not seeing why that check would make the error more principled (or at least, any more principled than what is on `main` today)?

File: modules/internal/ChapelArray.chpl
Comment (bradcray) [reply modules/internal/ChapelArray.chpl:331]: I don't know what primitive assignment will or won't handle, and just used it because it was what was used here before my changes, and worked for the int/real case, as you note.  I'm open to adding more testing, but what sort of non-trivial elements would you like me to test it with?

File: modules/internal/ChapelArray.chpl
Comment (bradcray) [reply modules/internal/ChapelArray.chpl:335]: I agree that it shouldn't happen, but was worried about users getting a worse error message if anything went wrong down the road (say, return type unification got smarter faster than `=` between the types did).  So I view this as an assertion that has a nicer user-facing form rather than an error we anticipate getting triggered.

File: modules/internal/ChapelArray.chpl
Comment (bradcray) [reply modules/internal/ChapelArray.chpl:312]: It definitely shouldn't have to in the current code organization.  I made it handle it anyway just for generality (rather than adding a where clause only making the routine apply to het. tuples, say), and in case it saw additional uses in the future (though I hope dyno will replace this functionality before that happens).

**Comment by vasslitvinov** [comment]
@bradcray looks good to merge. Let me follow up on the prim_assign issue post-merge.

I am planning to test these scenarios:
* owned classes, perhaps a parent class and unrelated children
* POD records of different sizes with implicit conversions defined between them

Also - any docs that need to be updated for this change?

**Comment by bradcray** [comment]
> owned classes, perhaps a parent class and unrelated children

Note that I don't think our return type unification handles parent/child cases today, based on https://github.com/chapel-lang/chapel/blob/93523cb4131786052f2b2e08b958ae2b8376982e/test/arrays/bradc/inferArrayOfDerivedType.chpl (which is why I didn't write more tests with classes).

> POD records of different sizes with implicit conversions defined between them

We don't support user-defined coercions today, so this is another case that's difficult to test for.

> Also - any docs that need to be updated for this change?

Excellent catch.  Yes there are, and I'll add that before merging.

**Comment by vasslitvinov** [comment]
inferArrayOfDerivedType.chpl tests an array literal with two sibling classes. Instead I wanted to test this:

```chpl
class C { }
class D: C { }
class E: C { }

var A = [new owned E(), new owned D(), new owned C()];
```

**Comment by bradcray** [comment]
> I can add that test to my PR if you like (whether it works or not).

Actually, I was too curious not to try it (as you may already know, it works), so will add this as another case to `hetLiteralsUnifiable.chpl`.

**Comment by bradcray** [comment]
> > Also - any docs that need to be updated for this change?
> 
> Excellent catch. Yes there are, and I'll add that before merging.

@vasslitvinov : I know you've approved this for merging, but if you'd be up for reading the docs commit (https://github.com/chapel-lang/chapel/pull/20563/commits/9fbda3d3cbd08d034724a27731e9ec41cf9ace49) before I do so, I'd appreciate a second pair of eyes on it.

**Comment by bradcray** [comment]
> I wanted to take out "representing" and "values" as they are a distraction in "representing the array elements' values", however did not find a way to do so.

If you prefer, we could say:

> Rectangular array literals are specified by enclosing a comma-separated list of expressions in square brackets, where the expressions represent the values of the array's elements.

(and maybe move the sentence about the trailing comma up just after this as well—it feels a bit like an afterthought in its current location).  Let me know if you prefer that.

**Comment by bradcray** [comment]
> Turns out that yes there are user-defined coercions between records, defined by init= -- see the third paragraph [in the spec](https://chapel-lang.org/docs/master/language/spec/conversions.html#implicit-conversions-for-initialization-and-assignment). init= requires also defining a cast.

Though it's that spec text as clear as it might be, to me, it seems to be trying to describe the initialization and assignment contexts only, based on the paragraph that precedes it.  Is that how you interpret it as well?  Whereas when I think of coercions / implicit conversions most generally, I think of:

```chapel
// using your definitions of R1 and R2 and the init= and :

proc foo(r: R1) {
  writeln("In foo, got: ", r);
}

foo(new R2());
```

which doesn't work.  Even changing the intent of `r` to `in` (and the intent of your `init=` to `in` as well) doesn't seem to work, though I think it's probably required (to avoid a ref type mismatch) and I thought it might since that would arguably make this into a form of initialization under the new interpretation of `in` (?).

Anyway, if that case doesn't work, it doesn't particularly surprise me that return type unification, written so long ago, couldn't handle this case:

```chapel
config const coinflip = true;

proc bar() {
  if coinflip then
    return new R1();
  else
    return new R2();
}

var r = bar();
```

though again, I could see how perhaps it might since it's arguably a form of initialization-via-returned-value.  I'm tagging @mppf on this to see how expected or surprising these behaviors are based on his revamped in/out work, or whether I'm missing something more subtle.

I don't personally feel particularly compelled to add a future for this case, because to me it falls into the broader topic of "we don't really have first-class support for user-defined coercions" yet, and I don't know that it adds value to add a bunch of test cases that showing lots of cases where we don't have them until we start claiming to more than we do.

(I should also add that the other reason I imagine we don't have them is because I ask for them every once in awhile and/or ask Michael how much work they would be, and the response is never "we've already got them, silly!")

**Comment by vasslitvinov** [comment]
My proposed future is surely orthogonal to the main push of this PR, so can be added later.

We need to be aware that Chapel no longer has "coercions / implicit conversions most generally". User-defined record coercions are supported for returns, which is specified in the bullet list under "initialization and assignment". Perhaps we should rename that subsection to "Initialization, assignment, and returns". Record coercions are not supported upon argument passing. We switched to this setup guided by user feedback.

> Rectangular array literals are specified by enclosing a comma-separated list of expressions in square brackets, where the expressions represent the values of the array's elements.

How about: Rectangular array literals are specified by enclosing a comma-separated list of array elements in square brackets. [Trailing comma goes here.] Each array element is an arbitrary expression.

Your version works for me as well.

**Comment by vasslitvinov** [comment]
Tada!

C++ provides [`std::common_type(T...)`](https://en.cppreference.com/w/cpp/types/common_type) that gives the type that all argument types can be coerced to. It makes sense for us to provide the same functionality in the Types module. We will need two flavors, one for assign/init/return conversions, the other for argument conversions. We could also start with the latter as it does not depend on user-defined coercions.

**Comment by mppf** [comment]
Responding to https://github.com/chapel-lang/chapel/pull/20563#issuecomment-1254371528 --

Yes we have some initialization-y patterns that might be described as user-defined implicit conversion; however when I use that term unqualified, I am talking about the function case (vs the initialization case). In the spec we tried to lay out the different kinds of (implicit) conversions.

I do think it'd be reasonable for the return type unification to work if the types involved can be `init=`'d into each other. The reason we treat the function call case differently from initialization is that there is more risk that the implicit conversions can add to confusion (about which function is called, say).

**Comment by bradcray** [comment]
Rough timing information on this PR, using a developer implementation of Chapel (so, debugging is turned on in the compiler and is likely to bog all these times down a bit):

compilation time for array literal with 5060 elements:

| expr types | previously | with PR|
|-----------|-----------|--------|
| int-only      | 1:02            | 0:27     |
| int/real       |   error          | 1:30      |

This matches my expectation: homogeneous array literals are now faster (due to eliminating a `param` loop over the elements), while heterogeneous ones are now possible, but slower than the homogeneous case (even before the change), due to a second `param` loop over the tuple of elements representing the array elements.

**Comment by ghbrown** [comment]
If it's any consolation about the timings:
- I personally consider this a big step in a good direction (eliminated a common error, improved usability, etc.)
- anywhere I can think of where I'd be using arrays with that many elements they'd already be in standardized form and could be transformed into the right type by casting the whole array if needed, or I'd be using some other method to read them, or worse comes to worst I'll manually preprocess them.

Thanks for your work on this, all!

**Comment by bradcray** [comment]
I'm not needing any consolation, but thanks for the comments anyway, Gabriel!  Specifically, I think it's great that homogeneous array literals are now faster, and it's wonderful to be able to write some common heterogeneous array literal patterns at all, even if it does cost compile time.  And I think your point about machine-generated array literals being more amenable to being of consistent type is a good one.  Also, we'll expect this to get better over time as we move the logic out of the module code and into the compiler.

I was actually just jumping on here to post the timings with a non-developer build, which are a bit less painful:

compilation time for array literal with 5060 elements (release mode):

| expr types | previously | with PR|
|-----------|-----------|--------|
| int-only      | 0:24            | 0:11     |
| int/real       |   error          | 0:27      |

So that doesn't look nearly as painful.

**Comment by bradcray** [comment]
(oh, I meant to add:  I think that there's still a little clean-up to do in the intersection of this and the linear algebra module that I noticed, and I want to sanity check some of the examples that you put onto issue #8196 that served as incentive for me to look into this, but will put any comments about that on #8196 itself.)

**Comment by vasslitvinov** [comment]
Belated thought: we should report an "unimplemented" error when the computed element type is a domain or an array. At least think through what should/will happen when that is the case. E.x., the first (0-th) element's runtime type will become the array element's runtime type.

**Pull request merged**
